From Natural Language to SQL: Building and Tracking a Multi-Lingual Query Engine (#132)
mlflow/mlflow-website: website/blog/2024-12-04-from-natural-language-to-sql/index.md

<!-- truncate -->

## Introduction

SQL is an essential skill needed for managing and accessing data in relational databases and many users fail to possess the necessary skills that are required to write correct SQL queries which can hinder them from utilizing data effectively. NL2SQL or Natural Language to SQL systems help in solving this problem by providing a translation from Natural Language to SQL commands allowing non-technical people to interact with data: users can just ask questions in a natural language they are comfortable speaking and these systems will assist them in returning the appropriate information.

However, there are also a number of problems that remain when creating an NL2SQL system like semantic ambiguity, schema mapping or error handling and user feedback. Therefore, it is very important that while building such systems, we must put some guardrails instead of completely relying on LLM.

> In this blog post, we’ll walk you through the process of building and managing the lifecycle of a Multilingual Query Engine, encompassing both Natural Language to SQL generation and query execution.

Reviewer: I don't see much about lifecycle management in the post (it mentions lifecycle management a few times, but doesn't show much about how to use MLflow for that purpose). You might de-emphasize that but show a little more about how MLflow tracing gives visibility into the many different components of a setup like this.

I think it's really cool, looking at the gif at the end, how we can see the different nodes laid out in the article reflected in the trace. It might be interesting to show screenshots of that for each section, or at least call that out more clearly in that section—specifically, that the final graph can be a bit of a black box, might be challenging to debug, to figure out what is happening at each step, but tracing gives really clear visibility into that with the one line of code.

We’ll start by demonstrating how to leverage LangGraph’s capabilities to build a dynamic AI workflow. This workflow integrates OpenAI and external data sources, such as a Vector Store and an SQLite database, to process user input, perform safety checks, query databases, and generate meaningful responses.

> Throughout this post, we’ll leverage MLflow’s Models from Code feature to manage the lifecycle of the Multilingual Query Engine. This approach allows the AI workflow to be treated like a traditional ML model, enabling tracking, versioning, and deployment across various serving infrastructures.

Reviewer 2: Might want to do a doc-links directly to this page for the Models from Code reference :) https://mlflow.org/docs/latest/model/models-from-code.html

Author: Thank you for pointing that out! Added.

# Requisites

To set up and run this project, ensure the following **Python packages** are installed:
- faiss-cpu
- langchain
- langchain-core
- langchain-openai
- langgraph
- langchain-community
- pydantic
- typing_extensions
- python-dotenv

Additionally, an **MLflow Tracking Server** is required to log and manage experiments, models, and traces effectively. For local setup, refer to the official MLflow documentation for instructions on [configuring a simple MLflow Tracking Server](https://mlflow.org/docs/latest/tracking/server.html).

Finally, ensure that your OpenAI API key is saved within a .env file in the project directory. This allows the application to securely access the OpenAI services required for building the AI workflow. The .env file should include a line like:

```

OPENAI_API_KEY=your_openai_api_key

```

# Multilingual Query Engine using Langraph

> The Multilingual Query Engine leverages LangGraph’s advanced features to create a stateful, multi-agent, and cyclical graph architecture.

Author: Thank you for the suggestion! I've updated the section to include a brief explanation of LangGraph as an AI orchestration tool and outlined why it was chosen for the Multi-Lingual Query Engine. Let me know if this addresses your comment or if you’d like further clarification.

## key Langraph features:

1. **Stateful Architecture**: The engine maintains a dynamic snapshot of the graph’s execution status. This snapshot acts as a shared resource across nodes, enabling efficient decision-making and real-time updates at each node execution.

2. **Multi-Agent Design**: The AI Workflow includes multiple interactions with OpenAI and other external tools throughout the workflow.

3. **Cyclical Graph Structure**: The graph’s cyclical nature introduces a robust retry mechanism. This mechanism dynamically addresses failures by looping back to previous stages when needed, ensuring continuous graph execution. (Details of this mechanism will be discussed later.)

> ## AI Workflow Overview

Reviewer 2: ```suggestion

## AI Workflow Overview
```
Linting - header sections need spaces on either side

Author: Fixed all linting.

The Multilingual Query Engine’s advanced AI workflow is composed of interconnected nodes and edges, each representing a crucial stage:

1. **Translation Node**: Converts the user’s input into English.

2. **Safety Checks**: Ensures user input is free from toxic or inappropriate content and does not contain harmful SQL commands (e.g., DELETE, DROP).

3. **Database Schema Extraction**: Retrieves the schema of the target database to understand its structure and available data.

4. **Relevancy Validation**: Validates the user’s input against the database schema to ensure alignment with the database’s capabilities.

5. **SQL Query Generation**: Generates an SQL query based on the user’s input and the current database schema.

6. **SQL Query Validation**: Executes the SQL query in a rollback-safe environment to ensure its validity before running it.

7. **Dynamic State Evaluation**: Determines the next steps based on the current state. If the SQL query validation fails, it loops back to Stage 5 to regenerate the query.

8. **Query Execution and Result Retrieval**: Executes the SQL query and returns the results if it’s a SELECT statement.

> The retry mechanism is introduced in Stage 7, where the system dynamically evaluates the current graph state. Specifically, when the SQL query validation node (Stage 6) detects an issue, the state triggers a loop back to the SQL Generation node (Stage 5) for a new SQL Generation attempt (within a maximum of 3 attemps).

Reviewer: The intro—first paragraph or two—could use a very brief summary of this. "We will build a system that takes natural language input, such as X, from the user, validates it for safety, and generates correct SQL, informed by context about the database schema." A clear description of the task and an example of the final workflow the project will enable will help readers get their bearings right from the beginning.

Author: I've revised the introduction to include a concise summary of the task and workflow, along with a concrete example to help readers quickly get the purpose and functionality of the Multi-Lingual Query Engine. Please let me know if this aligns with your suggestion.

## Components

The Multilingual Query Engine interacts with several external components to transform Natural Language user inputs into SQL queries and execute them in a safe and robust manner. In this section, we will take a detailed look at the key AI Workflow components: OpenAI, Vector Store, SQLite Database, and SQL Generation Chain.

### OpenAI

OpenAI, more specifically gpt-4 language model plays a crucial role in multiple stages of the workflow, providing the intelligence required for:

1. **Translation**: Translates user input into English. If the text is already in English, it simply repeats the input.

2. **Safety Checks**: Analyzes user input to ensure that it does not contain toxic or inappropriate content.

3. **Relevancy Checks**: Evaluates whether the user's question is relevant given the database schema.

4. **SQL Generation**: Generates valid and executable SQL queries based on user input, SQL generation documentation, and the database schema.

Details on OpenAI implementation will be provided later on in the Node implementation section.
### FAISS Vector Store

> 

Reviewer 2: Might want to mention alternatives to an in-memory Vector Store for permanent / more scalable shared resource embeddings storage that can be shared across projects.

To build an effective Natural Language to SQL engine capable of generating accurate and executable SQL queries, we leverage Langchain's FAISS Vector Store feature. This setup allows the system to search and extract SQL query generation guidelines from W3Schools SQL documents previously stored in the Vector Database, enhancing the success of SQL query generation.

> #### Step 1: Load SQL Documentation

Reviewer 2: Let's make sure to leave a blank new line on either side of any heading section

The first step in creating a FAISS Vector Store with SQL query generation guidelines is to load SQL documentation from the [W3Schools SQL page](https://www.w3schools.com/sql/) using Langchain's RecursiveUrlLoader. This tool retrieves the documentation, allowing us to use it as a knowledge base for our engine.
#### Step 2: Split the Text into Manageable Chunks

> The loaded SQL documentation is a lengthy text, making it difficult to be effectively ingested by the LLM. To address this, the next step involves splitting the text into smaller, manageable chunks using Langchain's RecursiveCharacterTextSplitter. By splitting the text into chunks of 500 characters with a 50-character overlap, we ensure the AI has sufficient context while minimizing the risk of losing important information that spans across chunks. The split_text method applies this splitting process, storing the resulting pieces in a list called documents.

Reviewer 2: ```suggestion
The loaded SQL documentation is a lengthy text, making it difficult to be effectively ingested by the LLM. To address this, the next step involves splitting the text into smaller, manageable chunks using Langchain's RecursiveCharacterTextSplitter. By splitting the text into chunks of 500 characters with a 50-character overlap, we ensure the language model has sufficient context while minimizing the risk of losing important information that spans across chunks. The split_text method applies this splitting process, storing the resulting pieces in a list called 'documents'.
```

        # Load SQL documentation
        url = "https://www.w3schools.com/sql/"
        loader = RecursiveUrlLoader(
            url=url, max_depth=2, extractor=lambda x: Soup(x, "html.parser").text
        )
        docs = loader.load()

        # Split documents into chunks
        text_splitter = RecursiveCharacterTextSplitter(
            chunk_size=500,
            chunk_overlap=50,
            separators=["\n\n", "\n", ".", "!", "?", ",", " ", ""],
        )

        documents = []
        for doc in docs:
            splits = text_splitter.split_text(doc.page_content)
            for i, split in enumerate(splits):
                documents.append({
                    "content": split,
                    "metadata": {"source": doc.metadata["source"], "chunk": i}
                })

        # Compute embeddings and create vector store
        embedding_model = OpenAIEmbeddings()
        vector_store = FAISS.from_texts(
            [doc["content"] for doc in documents],
            embedding_model,
            metadatas=[doc["metadata"] for doc in documents]
        )

        # Save the vector store to disk
        vector_store.save_local(vector_store_dir)

>         print("Vector store created and saved to disk.")

Reviewer 2: Could we convert the print statements to `_logger.info()` statements instead to show curious readers how to avoid common linting issues in their code?

Author: Thanks for the suggestion! I’ve updated the code to replace the print statements with _logger.info().

    return vector_store
```

### SQLite Database

> The SQLite database is a key component of the Multilingual Query Engine, serving as the structured data repository that supports SQL query efficient generation, validation and execution by enabling:

Reviewer 2: It might be worth mentioning why this is chosen for this example. Personally, I'm a huge fan of the portability and performance of sqlite and it definitely has many uses far beyond just demonstrations of concepts. As a self-contained data storage layer for an application, it's phenomenal at what it does and can greatly simplify developers' lives who would otherwise assume that they need to spin up a MySQL / PostGres DB for something that a local disk DB would handle much better.

Author: I’ve added a section detailing why SQLite was chosen for this example. I highlighted its lightweight, portable, and self-contained design, which, along with its ease of use and strong performance, makes it an ideal choice for this application.

    conn.commit()

def setup_database():
    """Setup the database and return the connection."""
    db_file = 'data/database.db'
    if not os.path.exists('data'):
        os.makedirs('data')

    db_exists = os.path.exists(db_file)

    conn = create_connection(db_file)

    if not db_exists:
        print("Setting up the database...")
        create_tables(conn)
        populate_tables(conn)
    else:
        print("Database already exists. Skipping setup.")

    return conn
```

### SQL Generation Chain

The **SQL Generation Chain** (`sql_gen_chain`) is the backbone of automated SQL query generation in our workflow. This chain leverages LangChain's modular capabilities and OpenAI's advanced natural language processing to transform user questions into precise and executable SQL queries.

**Core Features**:

- **Prompt-Driven Generation**: Begins with a thoughtfully designed prompt that integrates the database schema and documentation snippets, ensuring queries are contextually accurate.

- **Structured Responses**: Delivers outputs in a predefined format, including:

- A **description** of the query's purpose.

- The corresponding **SQL code** ready for execution.

> - **Adaptable and Reliable**: Uses GPT-4 for robust, consistent query generation, minimizing manual effort and errors.

Reviewer 2: Should we update this to use `gpt-4o-mini`? The tool calling functionality with that LLM build is far superior to base gpt-4

Author: Great suggestion! I’ve updated the workflow to use gpt-4o-mini.

Here is the user's question:""",
            ),
            ("placeholder", "{messages}"),
        ]
    )

    # Initialize the OpenAI LLM
    llm = ChatOpenAI(temperature=0, model="gpt-4")

    # Create the code generation chain
    sql_gen_chain = sql_gen_prompt | llm.with_structured_output(SQLQuery)

    return sql_gen_chain
```

### **Node Descriptions**
#### **1. Translate Input**

The `translate_input` node translates user queries into English to standardize processing, ensuring all inputs are compatible with downstream nodes.

**Purpose:** Normalize user input to English for consistent processing.

**Process:**

- Extracts the latest user input from the state.

- Uses the `ChatOpenAI` model to translate the input or verify it’s already in English.

- Updates the state with the translated input.

**Key Considerations:**

- Maintains original intent and context during translation.

- Provides accurate results even for multilingual queries.

**Examples:**

- Input: _"Quantos pedidos foram realizados em Novembro?"_

- Translated: _"How many orders were made in November?"_

- Input: _"Combien de ventes avons-nous enregistrées en France ?"_

> - Translated: _"How many sales did we record in France?"_

Reviewer: Note early on that multilingual refers to taking natural-language inputs in multiple languages, not e.g. supporting multiple SQL dialects or something like that. I wasn't 100% sure until I got to here!

Author: Thanks for pointing this out, it makes a lot of sense! I've clarified early on that the term 'multi-lingual' refers specifically to the system's ability to process natural language inputs in multiple languages, not to supporting multiple SQL dialects. I hope this resolves any potential ambiguity.

**Code:**

```python
def translate_input(state: GraphState):
    print("---TRANSLATING INPUT---")
    messages = state["messages"]
    user_input = messages[-1][1]  # Get the latest user input

    # Translation prompt for the model
    translation_prompt = f"""
Translate the following text to English. If the text is already in English, repeat it exactly without any additional explanation.

Text:
{user_input}
"""
    # Call the OpenAI LLM to translate the text
    translated_response = llm.invoke(translation_prompt)
    translated_text = translated_response.content.strip()  # Access the 'content' attribute and strip any extra spaces
    state["translated_input"] = translated_text  # Save the translated input
    print(f"Translated Input: {translated_text}")

    return state

> ```

Reviewer: Does translating to english before translating to SQL improve performance? Or would it work just as well to translate from whatever the input language is to SQL? Worth motivating this step.

- **Response:** *"Query allowed."*

2. **Inappropriate Content:**

- **Input:** *"Show me orders where customers have names like 'John the Idiot';"*

- **Response:** *"Your query contains inappropriate content and cannot be processed."*

- **Input:** *"Find total sales by region."*

- **Response:** *"Input is safe to process."*

**Code:**

```python
def safety_check(state: GraphState):
    print("---PERFORMING SAFETY CHECK---")
    translated_input = state["translated_input"]
    messages = state["messages"]
    error = "no"

    # List of disallowed SQL operations (e.g., DELETE, DROP)
    disallowed_operations = ['CREATE', 'DELETE', 'DROP', 'INSERT', 'UPDATE', 'ALTER', 'TRUNCATE', 'EXEC', 'EXECUTE']
    pattern = re.compile(r'\b(' + '|'.join(disallowed_operations) + r')\b', re.IGNORECASE)

    # Check if the input contains disallowed SQL operations
    if pattern.search(translated_input):
        print("Input contains disallowed SQL operations. Halting the workflow.")
        error = "yes"
        messages += [("assistant", "Your query contains disallowed SQL operations and cannot be processed.")]
    else:
        # Check if the input contains inappropriate content

>         safety_prompt = f"""

Reviewer: curious about using pattern search on the natural language input—wouldn't checking for disallowed SQL operations make more sense after the SQL is generated?

i.e. what would happen if the natural language input is "please get rid of the customers table." It doesn't look like checks for disallowed operations are run again after the sql is generated, so might it be possible that the system would generate and run a drop table command as long as the user didn't explicitly say "drop table"?

**Examples:**

- Input SQL: _"SELECT name FROM customers WHERE city = 'New York';"_

- Validation: Query is valid.

- Input SQL: _"SELECT MONTH(date) AS month, SUM(total) AS total_sales FROM orders GROUP BY MONTH(date);"_

- Response: _"Your SQL query failed to execute: no such function: MONTH."_

**Code:**

```python
def sql_check(state: GraphState):
    print("---VALIDATING SQL QUERY---")
    messages = state["messages"]
    sql_solution = state["generation"]
    error = "no"

    sql_code = sql_solution.sql_code.strip()

    try:
        # Start a savepoint for the transaction
        conn.execute('SAVEPOINT sql_check;')
        # Attempt to execute the SQL query
        cursor.execute(sql_code)
        # Roll back to the savepoint to undo any changes
        conn.execute('ROLLBACK TO sql_check;')
        print("---SQL QUERY VALIDATION: SUCCESS---")
    except Exception as e:
        # Roll back in case of error
        conn.execute('ROLLBACK TO sql_check;')
        print("---SQL QUERY VALIDATION: FAILED---")
        print(f"Error: {e}")
        messages += [("user", f"Your SQL query failed to execute: {e}")]
        error = "yes"

    state["error"] = error

    return state

> ```

Reviewer: This seems to validate that the SQL runs, but I don't see where it checks for safety or detects potentially destructive operations. As far as I can tell, a drop table command would clear this step. See earlier note—it looks like the safety check occurs before the sql is generated.

    sql_code = sql_solution.sql_code.strip()
    results = None
    no_records_found = False  # Flag to indicate no records found

    try:
        # Execute the SQL query
        cursor.execute(sql_code)
        # For SELECT queries, fetch and store results
        if sql_code.upper().startswith("SELECT"):
            results = cursor.fetchall()
            if not results:
                no_records_found = True
                print("---SQL QUERY EXECUTION: SUCCESS---")
                print("No records found.")
            else:
                print("---SQL QUERY EXECUTION: SUCCESS---")
        else:
            # For non-SELECT queries, commit the changes
            conn.commit()
            print("---SQL QUERY EXECUTION: SUCCESS---")
    except Exception as e:
        print("---SQL QUERY EXECUTION: FAILED---")
        print(f"Error: {e}")

    # Store the results and no_records_found flag in the state
    state["results"] = results
    state["no_records_found"] = no_records_found
    return state
```

The schema above provides a representation of the LangGraph Nodes and Edges:

![graph_representation](graph.png)

# Logging the Model in MLFlow

> Now that we have built a Multi-Lingual Query Engine using LangGraph, we are ready to log the model using MLflow’s [ Model from Code](https://mlflow.org/blog/models_from_code). This approach, where we log the code that represents the model, contrasts with object-based logging, where a model object is created, serialized, and logged as a pickle or JSON object.

Reviewer: Motivate this step (logging the model) a little more? e.g. can mention versioning, sharing, packaging for deployment, etc.

Author: I've expanded the section to better motivate the model logging step, emphasizing how MLflow enables tracking, versioning, and packaging for deployment across different serving infrastructures.

            "results": None,
            "generation": None,
            "no_records_found": False,
            "translated_input": "",  # Initialize translated_input
        }

        solution = app.invoke(initial_state)

        # Check if an error was set during the safety check
        if solution["error"] == "yes":
            print("\nAssistant Message:\n")
            print(solution["messages"][-1][1])  # Display the assistant's message
            continue  # Skip to the next iteration

        # Extract the generated SQL query from solution["generation"]
        sql_query = solution["generation"].sql_code
        print("\nGenerated SQL Query:\n")
        print(sql_query)

        # Extract and display the query results
        if solution.get("no_records_found"):
            print("\nNo records found matching your query.")
        elif "results" in solution and solution["results"] is not None:
            print("\nQuery Results:\n")
            for row in solution["results"]:
                print(row)
        else:
            print("\nNo results returned or query did not execute successfully.")

    print("Goodbye!")

if __name__ == "__main__":

>     main()

Reviewer: Can you show an example invocation of this (not all the code, just one quick example) toward the beginning?

Author: I've included an example invocation early on to help illustrate the workflow. The example demonstrates how a user input like 'Quantos clientes temos por país?' (in Portuguese) is processed step by step, from translation to SQL generation and execution.

With LangChain autologging, traces are automatically logged to the active MLflow experiment whenever invocation APIs are called on chains. This seamless integration ensures that every interaction is captured for monitoring and analysis.

## Viewing Traces in MLflow

Traces can be easily accessed by navigating to the MLflow experiment of interest and clicking on the "Tracing" tab. Once inside, selecting a specific trace provides detailed execution information.

Each trace includes:

1. **Execution Graphs**: Visualizations of the workflow steps.
2. **Inputs and Outputs**: Detailed logs of data processed at each step.

This granular visibility enables developers to debug and optimize their workflows effectively.

By leveraging MLflow tracing, we ensure that our Multi-Lingual Query Engine remains transparent, auditable, and scalable.

![mlflow_tracing_gif](mlflow_trace.gif)

> 

Reviewer: I really like this, would love some more emphasis on how tracing lets you visualize the whole graph execution.

Author: I’ve added more emphasis on how MLflow Tracing provides a comprehensive view of the entire graph execution.

We’ll start by demonstrating how to leverage [LangGraph’s](https://www.langchain.com/langgraph) capabilities to build a dynamic AI workflow. This workflow integrates OpenAI and external data sources, such as a Vector Store and an SQLite database, to process user input, perform safety checks, query databases, and generate meaningful responses.

Throughout this post, we’ll leverage [MLflow’s Models from Code](https://mlflow.org/docs/latest/model/models-from-code.html) feature to enable seamless tracking and versioning of AI Workflows. Additionally, we’ll deep dive [MLflow’s Tracing](https://mlflow.org/docs/latest/llms/tracing/index.html) feature, designed to enhance the observability of the many different components of an AI workflow by tracking inputs, outputs, and metadata associated with each intermediate step. This enables easy identification of bugs and unexpected behaviors, providing greater transparency over the workflow.

# Requisites

To set up and run this project, ensure the following **Python packages** are installed:
- `faiss-cpu`
- `langchain`
- `langchain-core`
- `langchain-openai`
- `langgraph`
- `langchain-community`

> - `pydantic`

Author: You are right, updated with the reference to the pydantic version.

### OpenAI

OpenAI, more specifically `gpt-4o-mini` language model plays a crucial role in multiple stages of the workflow, providing the intelligence required for:

1. **Translation**: Translates user input into English. If the text is already in English, it simply repeats the input.

2. **Safety Checks**: Analyzes user input to ensure that it does not contain toxic or inappropriate content.

3. **Relevancy Checks**: Evaluates whether the user's question is relevant given the database schema.

4. **SQL Generation**: Generates valid and executable SQL queries based on user input, SQL generation documentation, and the database schema.

Details on OpenAI implementation will be provided later on in the [Node Descriptions](#node-descriptions) section.

### FAISS Vector Store

To build an effective natural language to SQL engine capable of generating accurate and executable SQL queries, we leverage Langchain's FAISS Vector Store feature. This setup allows the system to search and extract SQL query generation guidelines from [W3Schools SQL documents](https://www.w3schools.com/sql/) previously stored in the Vector Database, enhancing the success of SQL query generation.

> For demo purposes, we are using FAISS, an in-memory vector store where vectors are stored directly in RAM. This provides fast access but means data is not persisted between runs. For a more scalable solution that enables embeddings to be stored and shared across multiple projects, we recommend alternatives like [AWS OpenSearch](https://aws.amazon.com/what-is/opensearch/), a cloud-based solution offering persistent storage, automatic scaling, and seamless integration with other cloud services, making it well-suited for large-scale applications.

Author: I've updated the text to include references to GCP's Vertex AI Vector Search, Azure Vector Search, and Databricks' Vector Search, along with their respective documentation links.

```python
import mlflow
from definitions import (
    EXPERIMENT_NAME,
    MODEL_ALIAS,
    REGISTERED_MODEL_NAME,
    REMOTE_SERVER_URI,
)
from mlflow import MlflowClient

client = MlflowClient(tracking_uri=REMOTE_SERVER_URI)

mlflow.set_tracking_uri(REMOTE_SERVER_URI)
mlflow.set_experiment(EXPERIMENT_NAME)

with mlflow.start_run():
    logged_model_info = mlflow.pyfunc.log_model(
        python_model="sql_model.py",
        artifact_path="sql_generator",
        registered_model_name=REGISTERED_MODEL_NAME,
        code_paths=["workflow.py"],
    )

client.set_registered_model_alias(
    REGISTERED_MODEL_NAME, MODEL_ALIAS, logged_model_info.registered_model_version
)

print(f"MLflow Run: {logged_model_info.run_id}")
print(f"Model URI: {logged_model_info.model_uri}")
```

> 

Reviewer 2: Might want to add a reference to validating the model's input schema as a check prior to registering utilizing the `mlflow.models.predict()` API (in master branch we now support `"uv"` as a backend environment for the subprocess validation of the model environment which significantly (10-100x speed improvement) improves the dependency fetch and pyenv creation for validation. We're releasing this feature next week as part of MLflow 2.20.0

Author: While I’ve gone through the MLflow 2.20.0 documentation, I didn't understand your suggestion here. Could you elaborate a bit further on how you envision incorporating this into the blog.

Reviewer 2: This was the doc that I was referring to: https://mlflow.org/docs/latest/python_api/mlflow.models.html#mlflow.models.predict 
In the example within that section we show the usage of the uv env_manager. 
The suggestion I posted was to mention that this API exists and can be used to save some potential future headaches during deployment. 

In your code, we're doing:

1. Create model definition script sql_model.py
2. Log & Register the model

My comment / suggestion would be something more like:

1. Create model definition script
2. Log the model as code (but don't register)
3. Validate with mlflow.models.predict on the logged model's uri using uv as env_manager
4. Register the model

This extra validation step and confirmation that the dependencies are set correctly and that the interface signature is correct can save hours of frustrating debugging and fixes after deploying a potentially broken model.

Author: Hello @BenWilson2,

I tried to do the changes but it gives me an error regarding input data validation. I checked the error (cannot pickle 'sqlite3.Connection' object) and I have found: 

"An SQLite connection is a stateful object synchronized with your system via the IOs it performed. By SQLite's library design, it will not be able to support the duplication and the sharing of the underlying handles of system resources."

This means that deepcopy is not something we can do when the input is an SQLite connection.
![Uploading Screenshot 2025-01-23 at 15.11.56.png…]()

In the MLflow UI, the stored model includes both the `sql_model.py` and `workflow.py` scripts as artifacts within the run. This logging from code feature not only records the model's parameters and metrics but also captures the code defining its functionality. This ensures observability, seamless tracking, and straightforward debugging directly through the UI.

> 

Reviewer 2: Please add a warning note to users to ensure that sensitive elements like API Keys should never be defined in code being used for models from code logging as token leakage can occur since we store everything in the script definition as-is.

Author: I’ve added a warning to explicitly inform users that any tokens or credentials hardcoded in the scripts will be logged as part of the script in the MLflow artifact.

---
title: "From Natural Language to SQL: Building and Tracking a Multi-Lingual Query Engine"
description: MLflow Models from Code and MLflow Tracing applied to AI Workflows
tags:
  - pyfunc
  - mlflow
  - sql-generator
  - models-from-code
  - tracing
slug: from-natural-language-to-sql
authors:
  - hugo-carvalho
  - joana-ferreira
  - rahul-pandey

> thumbnail: img/blog/from-natural-language-to-sql.png

Reviewer 2: ```suggestion
thumbnail: /img/blog/from-natural-language-to-sql.png
```

        }

        solution = app.invoke(initial_state)

        # Check if an error was set during the safety check
        if solution["error"] == "yes":
            _logger.info("\nAssistant Message:\n")
            _logger.info(solution["messages"][-1][1])  # Display the assistant's message
            continue  # Skip to the next iteration

        # Extract the generated SQL query from solution["generation"]
        sql_query = solution["generation"].sql_code
        _logger.info("\nGenerated SQL Query:\n")
        _logger.info(sql_query)

        # Extract and display the query results
        if solution.get("no_records_found"):
            _logger.info("\nNo records found matching your query.")
        elif "results" in solution and solution["results"] is not None:
            _logger.info("\nQuery Results:\n")
            for row in solution["results"]:
                _logger.info(row)
        else:
            _logger.info("\nNo results returned or query did not execute successfully.")

    _logger.info("Goodbye!")

if __name__ == "__main__":
    main()

```

# Project File Structure

The Project follows a simple file structure:

> <img src="file_structure.png" alt="file_structure" width="300">

Reviewer 2: ```suggestion
<img src={require("./file_structure.png").default} alt="file structure" width="60%"/>
```

Reviewer 2: This is required for the resizing and the proper rendering

With LangChain autologging, traces are automatically logged to the active MLflow experiment whenever invocation APIs are called on chains. This seamless integration ensures that every interaction is captured for monitoring and analysis.

## Viewing Traces in MLflow

Traces can be easily accessed by navigating to the MLflow experiment of interest and clicking on the "Tracing" tab. Once inside, selecting a specific trace provides detailed execution information.

Each trace includes:

1. **Execution Graphs**: Visualizations of the workflow steps.
2. **Inputs and Outputs**: Detailed logs of data processed at each step.

By leveraging MLflow Tracing, we gain granular visibility into the whole graph execution. AI workflows graphs can often feel like a black box, making it challenging to debug and understand what is happening at each step. However, with just a single line of code to enable tracing, MLflow provides clear and detailed insights into the workflow, allowing developers to effectively debug, monitor, and optimize every node of the graph ensuring that our Multi-Lingual Query Engine remains transparent, auditable, and scalable.

> ![mlflow_tracing_gif](mlflow_trace.gif)

Reviewer 2: This .gif doesn't quite render properly. Did you use ffmpeg to generate the .gif? If you used a different tool to convert to a .gif format (such as an online tool or other program), the encoding is likely not explicitly compatible with site renderers in REACT. 
Could you re-encode this file using 
```
ffmpeg -i <my_movie>.mp4 -vf "fps=10,scale=2048:-1:flags=lanczos,split[s0][s1];[s0]palettegen=max_colors=256:stats_mode=diff[p];[s1][p]paletteuse=dither=bayer:bayer_scale=5" <my_output_gif>.gif
```
And then replace the .gif to ensure that it renders locally in a local webserver?
